pyface/data_view/tests/test_data_view_widget.py

```diff
@@ -185,6 +502,52 @@ def test_selection_mode_extended(self):
             [((1, 4), ()), ((2, 0), ())],
         )
 
+    def test_selection_mode_change_when_selection_change(self):
+        modes = cycle(["extended", "single"])
+
+        self.widget.selection_mode = next(modes)
+        self._create_widget_control()
+        self.assertEqual(self.widget._get_control_selection_mode(), "extended")
+
+        def change_selection_mode(event):
+            if self.widget.selection:
+                self.widget.selection_mode = next(modes)
+
+        self.widget.observe(
+            change_selection_mode, "selection.items", dispatch="ui")
+        self.addCleanup(
+            self.widget.observe,
+            change_selection_mode, "selection.items", dispatch="ui",
+            remove=True,
+        )
+
+        # switch from "extended" to "single"
+        self.widget.selection = [((0,), ()), ((1,), ())]
+        self.gui.process_events()
+        self.assertEqual(self.widget._get_control_selection_mode(), "single")
```

Comment:
Just to clarify, this is not how selection mode works! Selection mode tells you what sorts of things can be selected, and does not change based on what the actual selection _is_.

When selection mode is `single` then the selection is always a list with one item.

When selection mode is `extened` then the selection is a list with any number of items, _including_ possibly just one item.

Reply:
I am confused. From reading the code, the widget resets the selection to an empty list whenever selection mode is changed.
The line you are commenting on is asserting the selection *mode*. When the selection mode is changed on the `self.widget`, the selection mode on the GUI control should be synchronized to the new value, right?

Reply:
To clarify, this test is testing the logic in this method:
https://github.com/enthought/pyface/blob/9daeb87e832e85d4ce0fbd29a5dc1a700bdec3bb/pyface/ui/qt4/data_view/data_view_widget.py#L91-L95